src/main/java/com/microsoft/sqlserver/jdbc/SQLServerSecurityUtility.java

```diff
@@ -215,4 +234,164 @@ static void verifyColumnMasterKeyMetadata(SQLServerConnection connection, String
+
+                    BufferedReader reader = new BufferedReader(new InputStreamReader(stream, UTF_8), 100);
+                    String result = reader.readLine();
+
+                    int startIndex_AT = result.indexOf(ActiveDirectoryAuthentication.ACCESS_TOKEN_IDENTIFIER)
+                            + ActiveDirectoryAuthentication.ACCESS_TOKEN_IDENTIFIER.length();
+
+                    String accessToken = result.substring(startIndex_AT, result.indexOf("\"", startIndex_AT + 1));
+
+                    Calendar cal = new Calendar.Builder().setInstant(new Date()).build();
+
+                    if (isAzureFunction) {
+                        // Fetch expires_on
+                        int startIndex_ATX = result
+                                .indexOf(ActiveDirectoryAuthentication.ACCESS_TOKEN_EXPIRES_ON_IDENTIFIER)
+                                + ActiveDirectoryAuthentication.ACCESS_TOKEN_EXPIRES_ON_IDENTIFIER.length();
+                        String accessTokenExpiry = result.substring(startIndex_ATX,
+                                result.indexOf("\"", startIndex_ATX + 1));
+                        if (connectionlogger.isLoggable(Level.FINER)) {
+                            connectionlogger.finer("MSI auth token expires on: " + accessTokenExpiry);
+                        }
+
+                        DateFormat df = new SimpleDateFormat(
+                                ActiveDirectoryAuthentication.ACCESS_TOKEN_EXPIRES_ON_DATE_FORMAT);
+                        cal = new Calendar.Builder().setInstant(df.parse(accessTokenExpiry)).build();
+                    } else {
+                        // Fetch expires_in
+                        int startIndex_ATX = result
+                                .indexOf(ActiveDirectoryAuthentication.ACCESS_TOKEN_EXPIRES_IN_IDENTIFIER)
+                                + ActiveDirectoryAuthentication.ACCESS_TOKEN_EXPIRES_IN_IDENTIFIER.length();
+                        String accessTokenExpiry = result.substring(startIndex_ATX,
+                                result.indexOf("\"", startIndex_ATX + 1));
+                        cal.add(Calendar.SECOND, Integer.parseInt(accessTokenExpiry));
+                    }
+
+                    return new SqlFedAuthToken(accessToken, cal.getTime());
+                }
+            } catch (Exception e) {
+                retry++;
+                // Below code applicable only when !isAzureFunctcion (VM)
+                if (retry > maxRetry) {
+                    // Do not retry if maxRetry limit has been reached.
+                    break;
+                } else {
+                    try {
+                        int responseCode = connection.getResponseCode();
+                        // Check Error Response Code from Connection
+                        if (410 == responseCode || 429 == responseCode || 404 == responseCode
+                                || (500 <= responseCode && 599 >= responseCode)) {

```

Comment:
Response codes shouldn't be magic numbers, better to create constants and document a little of their meaning in comments. I think this part of the code is just moved though so please document/create a work item.